go-controller/pkg/node/ovn_db_mgmt.go

```diff
@@ -0,0 +1,139 @@
+package node
+
+import (
+	"regexp"
+	"strings"
+	"time"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/metrics"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+func (n *OvnNode) ensureOvnRaft(db string) {
+	ticker := time.NewTicker(60 * time.Second)
+	klog.Infof("Starting ensure routine for Raft db: %s", db)
+	kclient := n.Kube.(*kube.Kube)
+	for {
+		select {
+		case <-ticker.C:
+			if dbOnNode, _ := metrics.CheckPodRunsOnGivenNode(kclient.KClient.(*kubernetes.Clientset),
+				"name=ovnkube-db", n.name, false); dbOnNode {
+				klog.Infof("Node %s has a OVN DB pod, ensuring OVN Raft membership", n.name)
+				n.ensureLocalRaftServerID(db)
+				n.ensureClusterRaftMembership(db)
+			}
+		case <-n.stopChan:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+// ensureLocalRaftServerID is used to ensure there is no stale member in the Raft cluster with our address
+func (n *OvnNode) ensureLocalRaftServerID(db string) {
+	var dbName string
+	var appCtl func(args ...string) (string, string, error)
+
+	if strings.Contains(db, "ovnnb") {
+		dbName = "OVN_Northbound"
+		appCtl = util.RunOVNNBAppCtl
+	} else {
+		dbName = "OVN_Southbound"
+		appCtl = util.RunOVNSBAppCtl
+	}
+
+	out, stderr, err := util.RunOVSDBTool("db-sid", db)
+	if err != nil {
+		klog.Warningf("Unable to get db server ID for: %s, stderr: %v, err: %v", db, stderr, err)
+		return
+	}
+	if len(out) < 4 {
+		klog.Errorf("Invalid db id found: %s for db: %s", out, db)
+		return
+	}
+	// server ID in raft membership is only first 4 char prefix
+	serverID := out[:4]
+	out, stderr, err = appCtl("cluster/status", dbName)
+	if err != nil {
+		klog.Warningf("Unable to get cluster status for: %s, stderr: %v, err: %v", db, stderr, err)
+		return
+	}
+	r, _ := regexp.Compile(`Address: *((ssl|tcp):[?[a-z0-9.:]+]?)`)
+	matches := r.FindStringSubmatch(out)
```

Comment:
There are easier ways to get the Server ID from cluster/status output, so  I don't know why we need to start from `Address: `. In the output below , the very first line is Server ID and there is a column for 'Server ID' itself.


```
]# ovs-appctl -t /var/run/openvswitch/ovnnb_db.ctl cluster/status OVN_Northbound
42db    <============== **SERVER ID**
Name: OVN_Northbound 
Cluster ID: d92f (d92fa139-71be-4beb-b361-0ec571d75965)
Server ID: 42db (42dbe4b3-9a94-4569-93fd-ee3dd17bdc12) <========= ** SERVER ID **
Address: tcp:10.0.16.14:6643
Status: cluster member
Role: follower
Term: 75
Leader: 71f8
Vote: 71f8

Election timer: 5000
Log: [87261, 88800]
Entries not yet committed: 0
Entries not yet applied: 0
Connections: ->7e36 ->71f8 <-7e36 <-71f8
Servers:
    7e36 (7e36 at tcp:10.0.16.29:6643)
    71f8 (71f8 at tcp:10.0.2.20:6643)
    42db (42db at tcp:10.0.16.14:6643) (self)
```

Reply:
In this particular line im getting the address. I get the server id above on line 50: `out, stderr, err := util.RunOVSDBTool("db-sid", db)`

Reply:
So, you get the ServerID from the DB file directly using `ovsdb-tool db-sid` and then you get the ServerID from the running daemon, and you do compare.

I am talking about getting the ServerID from the running daemon. Why can't we just take the value from the 1st line in the `cluster/status` output.

Reply:
We can. I thought sb-id might be more authoritative.